### ISSUE TITLE
PNP not entirely functional

### ISSUE THREAD

**Issue by ma-sadeghi** [PMEAL/OpenPNM#1380]
Just for future reference: Here's how to reproduce the bug:
1. Edit `openpnm/algorithms/TransientReactiveTransport.py`
2. Go to `_t_run_reactive` method, edit the following snippet from
```
            self._apply_sources()
            self._correct_apply_sources()
            # Compute the normalized residual
            res = np.linalg.norm(self.b-self.A*x)/ref
            if res >= self.settings['rxn_tolerance']:
                logger.info('Tolerance not met: ' + str(res))
                x_new = self._solve()
```
to
```
            self._apply_sources()
            self._correct_apply_sources()
            # Compute the normalized residual
            res = np.linalg.norm(self.b-self.A*x)/ref
            ##########
            if self._A.data.mean() == 0:
                raise Exception("A is empty!")
            ##########
            if res >= self.settings['rxn_tolerance']:
                logger.info('Tolerance not met: ' + str(res))
                x_new = self._solve()
```
3. Run `/scripts/example_transient_PNP.py`
4. See the output 
```
――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
Running TransientIonicTransport

Current time step: 100 s
Algorithms: alg_02, alg_03, alg_04
Time residuals: 1000000.0, 1000000.0, 1000000.0
Gummel iter: 1, residuals: 1000000.0, 1000000.0, 1000000.0
Traceback (most recent call last):

  File "/home/amin/Code/OpenPNM/scripts/example_transient_PNP.py", line 99, in <module>
    pnp.run()

  File "/home/amin/Code/OpenPNM/openpnm/algorithms/TransientIonicTransport.py", line 118, in run
    self._run_transient(t=t)

  File "/home/amin/Code/OpenPNM/openpnm/algorithms/TransientIonicTransport.py", line 217, in _run_transient
    p_alg._t_run_reactive(x0=i_old[p_alg.name])

  File "/home/amin/Code/OpenPNM/openpnm/algorithms/TransientReactiveTransport.py", line 396, in _t_run_reactive
    raise Exception("A is empty!")

Exception: A is empty!
```
Now, the reason this (possible) bug was not caught by the solver is a direct consequence of the way stopping criterion has been implemented, which is:
```
ref = np.sum(np.absolute(self._A_t.diagonal())) or 1
res = np.linalg.norm(self.b-self.A*x)/ref
if res >= self.settings['rxn_tolerance']:
    THE_REST_OF_THE_CODE
```
When A is empty, the `ref` gets the value of `1`, meaning that we're normalizing `res` by a very large number, and therefore, the algorithm mistakenly assumes convergence has been reached, whereas in reality, `A` is not even valid.

**Comment by jgostick** [comment]
@ma-sadeghi try to reproduce with dev....if not then close.

**Comment by ma-sadeghi** [comment]
I can't  reproduce it anymore, possibly fixed in PR #1451.
